endpoint/endpoint.go

```diff
@@ -276,6 +292,47 @@ func (e *Endpoint) GetLoggerConfig() log.LoggerConfig {
 	return e.config.LoggerConfig
 }
 
+// CreateOrUpdateMetrics uses DataRegistryService CreateOrUpdateMetrics to create or update metrics
+func (e *Endpoint) CreateOrUpdateMetrics(ctx context.Context, metricWrappers []*data_receiver.MetricWrapper) ([]*RegisterResponse, error) {
+	if e.config.APIKey != "" {
+		ctx = metadata.AppendToOutgoingContext(ctx, APIKeyHeader, e.config.APIKey)
+	}
+	stream, err := e.regclient.CreateOrUpdateMetrics(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, metricWrapper := range metricWrappers {
+		request := &zproto.RegisterMetricRequest{
+			Metric:     metricWrapper,
+			UpdateMode: zproto.UpdateMode_REPLACE, // replace by default
+		}
+		if err := stream.Send(request); err != nil {
+			log.Log(e, log.LevelError, log.Fields{
+				"metric wrapper": metricWrapper,
+			}, "Error sending metric wrapper to data-registry")
+			return nil, err
```

Comment:
Does the stream need to be explicitly closed or is the leak prevented automatically?

Reply:
I looked at some documentation online about the auto-generated code and what I could see was that we do expect to call the Send method within the loop and the CloseAndRecv just once when done. There's not a lot of standardization I can see around error handling (as in case of error you can't even be sure that the CloseAndRecv will work). Will look into this some more